python/ray/data/_internal/planner/filter.py

```diff
@@ -24,6 +24,9 @@ def fn(
             for row in block.iter_rows():
                 if row_fn(row):
                     builder.add(row)
-            return [builder.build()]
+            # NOTE: this yields an empty block if all rows are filtered out.
+            # This causes different behavior between filter and other map-like
+            # functions. We should revisit and try to get rid of this logic.
+            yield builder.build()
```

Comment:
Looks like if getting rid of this behavior (i.e. output empty block), it will cause CI tests failure ([link](https://buildkite.com/ray-project/oss-ci-build-pr/builds/11120#01860b76-66ae-4913-98fb-c6bcf0a4e42f)). So keep the original behavior here and we shall think a more proper fix in 2.4.